python/cudf/cudf/_lib/pylibcudf/groupby.pyx

```diff
@@ -0,0 +1,251 @@
+    aggregation_request,
+    aggregation_result,
+    groupby,
+    groups,
+    scan_request,
+)
+from cudf._lib.cpp.scalar.scalar cimport scalar
+from cudf._lib.cpp.table.table cimport table
+from cudf._lib.cpp.types cimport size_type
+
+from .aggregation cimport Aggregation
+from .column cimport Column
+from .table cimport Table
+from .types cimport null_policy, sorted
+from .utils cimport _as_vector
+
+
+cdef class GroupByRequest:
+    """A request for a groupby aggregation or scan.
+
+    Parameters
+    ----------
+    values : Column
+        The column to aggregate.
+    aggregations : List[Aggregation]
+        The list of aggregations to perform.
+    """
+    def __init__(self, Column values, list aggregations):
+        self._values = values
+        self._aggregations = aggregations
+
+    cdef aggregation_request _to_libcudf_agg_request(self) except *:
+        """Convert to a libcudf aggregation_request object.
+
+        This method is for internal use only. It creates a new libcudf
+        :cpp:class:`cudf::groupby::aggregation_request` object each time it is
+        called.
+        """
+        cdef aggregation_request c_obj
+        c_obj.values = self._values.view()
+
+        cdef Aggregation agg
+        for agg in self._aggregations:
+            c_obj.aggregations.push_back(move(agg.clone_underlying_as_groupby()))
+        return move(c_obj)
+
+    cdef scan_request _to_libcudf_scan_request(self) except *:
+        """Convert to a libcudf scan_request object.
+
+        This method is for internal use only. It creates a new libcudf
+        :cpp:class:`cudf::groupby::scan_request` object each time it is
+        called.
+        """
+        cdef scan_request c_obj
+        c_obj.values = self._values.view()
+
+        cdef Aggregation agg
+        for agg in self._aggregations:
+            c_obj.aggregations.push_back(move(agg.clone_underlying_as_groupby_scan()))
+        return move(c_obj)
+
+
+cdef class GroupBy:
+    """Group values by keys and compute various aggregate quantities.
+
+    Parameters
+    ----------
+    keys : Table
+        The columns to group by.
+    null_handling : null_policy, optional
+        Whether or not to include null rows in ``keys``. Default is null_policy.EXCLUDE.
+    keys_are_sorted : sorted, optional
+        Whether the keys are already sorted. Default is sorted.NO.
+    """
+    def __init__(
+        self,
+        Table keys,
+        null_policy null_handling=null_policy.EXCLUDE,
+        sorted keys_are_sorted=sorted.NO
+    ):
+        self.c_obj.reset(new groupby(keys.view(), null_handling, keys_are_sorted))
+
+    @staticmethod
+    cdef tuple _parse_outputs(
+        pair[unique_ptr[table], vector[aggregation_result]] c_res
+    ):
+        # Convert libcudf aggregation/scan outputs into pylibcudf objects.
+        # This function is for internal use only.
+        cdef Table group_keys = Table.from_libcudf(move(c_res.first))
+
+        cdef int i, j
```

Comment:
In pandas we typically use `Py_ssize_t` for loop variables, but that shouldn't make too much of a difference compared to `int`

Reply:
Yeah that is probably technically more correct, but agree it shouldn't matter much in our Cython code.